### PR TITLE
Add Clear Cache API for Searchable Snapshots

### PR DESCRIPTION
**Pull request by tlrx**
This pull request adds an API to clear the cache used for searchable snapshots:

```
POST /_searchable_snapshots/cache/clear
POST <index>/_searchable_snapshots/cache/clear
```

This API is useful in tests or to simply free disk space if needed. Note that this API does not clear cache stats.

Relates #50999

### CODE REVIEW

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/AbstractTransportSearchableSnapshotsAction.java [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        assert cacheDirectory != null;
+        assert cacheDirectory.getShardId().equals(shardRouting.shardId());
+        return executeShardOperation(request, shardRouting, cacheDirectory);
+    }
+
+    protected abstract ShardOperationResult executeShardOperation(Request request, ShardRouting shardRouting,
+                                                                  CacheDirectory cacheDirectory) throws IOException;
+
+    @Nullable
+    private static CacheDirectory unwrap(Directory dir) {
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/AbstractTransportSearchableSnapshotsAction.java]: Maybe `unwrapCacheDirectory` to be a bit more specific about how far we are unwrapping?

(I realise this code existed beforehand, but I only just thought of this)

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/rest/RestClearSearchableSnapshotsCacheAction.java [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    public List<Route> routes() {
+        return List.of(
+            new Route(RestRequest.Method.POST, "/_searchable_snapshots/cache/clear"),
+            new Route(RestRequest.Method.POST, "/{index}/_searchable_snapshots/cache/clear")
+        );
+    }
+
+    @Override
+    public String getName() {
+        return "clear_searchable_snapshots_indices_action";
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/rest/RestClearSearchableSnapshotsCacheAction.java]: I think we should have `cache` in the  name here somewhere. I tried a couple of different word orders and none was perfect. Maybe this?

```suggestion
        return "clear_indices_searchable_snapshots_cache_action";
```

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java [hunk 3]
```diff
@@ -61,7 +64,7 @@ private void runSearchableSnapshotsTest(SearchableSnapshotsTestCaseBody testCase
 
         logger.info("creating index [{}]", indexName);
         createIndex(indexName, Settings.builder()
-            .put(IndexMetaData.SETTING_NUMBER_OF_SHARDS, numberOfShards)
+            .put(SETTING_NUMBER_OF_SHARDS, numberOfShards)
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java]: nit: seems unnecessary,  and we use the fully-qualified  name below.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java [hunk 4]
```diff
@@ -162,6 +161,51 @@ public void testCloseAndReopen() throws Exception {
...
+            .filter(o -> o instanceof Map)
+            .map(o -> ((Map<?,?>)o).get("files"))
+            .filter(o -> o instanceof List)
+            .flatMap(o -> ((List) o).stream())
+            .filter(o -> o instanceof Map)
+            .map(o -> ((Map<?,?>)o).get("cached_bytes_written"))
+            .filter(o -> o instanceof Map)
+            .map(o -> ((Map<?,?>)o).get("sum"))
+            .mapToLong(o -> ((Number) o).longValue())
+            .sum();
```
Comment (DaveCTurner) [inline x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java]: I wonder whether it'd be better to do this whole test in the YAML suite. I think that's possible. WDYT?

**Review by DaveCTurner — APPROVED** [review]
LGTM. I left four very minor comments.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/AbstractTransportSearchableSnapshotsAction.java (hunk 1 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/action/AbstractTransportSearchableSnapshotsAction.java]: Sure, I pushed 544f0565c315c317585db9961cb654f5e81efcac

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/rest/RestClearSearchableSnapshotsCacheAction.java (hunk 2 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/xpack/searchablesnapshots/rest/RestClearSearchableSnapshotsCacheAction.java]: Thanks for spotting this. I pushed f1ce506f4605ebf0fe669db0a74e6c2e61fc765a to use your suggestion.

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java (hunk 3 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java]: I reverted this change in 244b89e1278f52e5ff2d1f7b3b550b7cafaf56f6

File: x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java (hunk 4 above)
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java]: I'm not sure I can do it in YAML: this test computes the total bytes written in cache (the stats API does not compute total among all shards and files) after a given query, then clear the cache and run the query again to verify that more data has been written in cache gain.

I pushed 587819a2c8cbdd0048c4c4c24a762367ea0e5969 to run the same query before/after the cache clearance.

Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/AbstractSearchableSnapshotsRestTestCase.java]: Oh yes so it does. Let's not add  `.sum()` to the YAML test runner :)

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/oss-distro-docs

**Comment by tlrx** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by tlrx** [comment]
@elasticmachine update branch

**Comment by tlrx** [comment]
Thanks David

**Pull request merged**